modules/040-node-manager/templates/node-group/_bootstrap.tpl

```diff
@@ -55,39 +41,153 @@ EOF
+      {{- if .proxy.noProxy }}
+  export NO_PROXY={{ .proxy.noProxy | join "," | quote }}
+  export no_proxy=${NO_PROXY}
+      {{- end }}
+    {{- else }}
+  unset HTTP_PROXY http_proxy HTTPS_PROXY https_proxy NO_PROXY no_proxy
+    {{- end }}
+  {{- end }}
+  {{- if $context.Values.nodeManager.internal.packagesProxy }}
+  export PACKAGES_PROXY_ADDRESSES="{{ $context.Values.nodeManager.internal.packagesProxy.addresses | join "," }}"
+  export PACKAGES_PROXY_TOKEN="{{ $context.Values.nodeManager.internal.packagesProxy.token }}"
+  {{- end }}
+
+  function check_python() {
+    for pybin in python3 python2 python; do
+      if command -v "$pybin" >/dev/null 2>&1; then
+        python_binary="$pybin"
+        return 0
+      fi
+    done
+    echo "Python not found, exiting..."
+    return 1
+  }
+
+  bb-package-install() {
+    local PACKAGE_WITH_DIGEST
+    for PACKAGE_WITH_DIGEST in "$@"; do
+      local PACKAGE=""
+      local DIGEST=""
+      PACKAGE="$(awk -F ":" '{print $1}' <<< "${PACKAGE_WITH_DIGEST}")"
+      DIGEST="$(awk -F ":" '{print $2":"$3}' <<< "${PACKAGE_WITH_DIGEST}")"
+      bb-package-fetch "${PACKAGE_WITH_DIGEST}"
+      local TMP_DIR=""
+      TMP_DIR="$(mktemp -d)"
+      tar -xf "${BB_FETCHED_PACKAGES_STORE}/${PACKAGE}/${DIGEST}.tar.gz" -C "${TMP_DIR}"
+
+      # shellcheck disable=SC2164
+      pushd "${TMP_DIR}" >/dev/null
+      ./install
+      popd >/dev/null
+      mkdir -p "${BB_INSTALLED_PACKAGES_STORE}/${PACKAGE}"
+      echo "${DIGEST}" > "${BB_INSTALLED_PACKAGES_STORE}/${PACKAGE}/digest"
+      cp "${TMP_DIR}/install" "${TMP_DIR}/uninstall" "${BB_INSTALLED_PACKAGES_STORE}/${PACKAGE}"
+      rm -rf "${TMP_DIR}" "${BB_FETCHED_PACKAGES_STORE:?}/${PACKAGE}"
+    done
+  }
+
+  bb-package-fetch() {
+    mkdir -p "${BB_FETCHED_PACKAGES_STORE}"
+    declare -A PACKAGES_MAP
+    local PACKAGE_WITH_DIGEST
+    for PACKAGE_WITH_DIGEST in "$@"; do
+      local PACKAGE=""
+      local DIGEST=""
+      PACKAGE="$(awk -F ":" '{print $1}' <<< "${PACKAGE_WITH_DIGEST}")"
+      DIGEST="$(awk -F ":" '{print $2":"$3}' <<< "${PACKAGE_WITH_DIGEST}")"
+      PACKAGES_MAP[$DIGEST]="${PACKAGE}"
+    done
+    bb-package-fetch-blobs PACKAGES_MAP
+  }
+
+  bb-package-fetch-blobs() {
+    local PACKAGE_DIGEST
+    for PACKAGE_DIGEST in "${!PACKAGES_MAP[@]}"; do
+      local PACKAGE_DIR="${BB_FETCHED_PACKAGES_STORE}/${PACKAGES_MAP[$PACKAGE_DIGEST]}"
+      mkdir -p "${PACKAGE_DIR}"
+      bb-package-fetch-blob "${PACKAGE_DIGEST}" "${PACKAGE_DIR}/${PACKAGE_DIGEST}.tar.gz"
+    done
+  }
+
+  bb-package-fetch-blob() {
+    check_python
+
+    cat - <<EOFILE | $python_binary
+import random
+import ssl
+try:
+  from urllib.request import urlopen, Request, HTTPError
+except ImportError:
+  from urllib2 import urlopen, Request, HTTPError
+endpoints = "${PACKAGES_PROXY_ADDRESSES}".split(",")
+# Choose a random endpoint as first ep, that increase fault tolerance and reduce load on first endpoint.
+endpoints.insert(0, random.choice(endpoints))
```

Comment:
Can we use `random.shuffle(endpoints)`?

Reply:
We can, but to improve fault tolerance we want to access the same first random endpoint twice. For example, in single master `random.shuffle(endpoints)` will return the same list with 1 try/1 endpoint, while `endpoints.insert(0, random.choice(endpoints))` will give two query attempts.

Reply:
I don't see much difference in these cases, other than adding extra conditions to code.

Reply:
Although, additional retries can cover cases when all rpp endpoints are unavailable at the time of bootstrap, and only 1 of them will be valid after some time. But I don’t know how real this case is.

Reply:
I think we should add retry logic. Logic with duplicated endpoints is non-intuitive.